### PR TITLE
Fix direct leaks in deterministic-unit_test_wavefunction_trialwf

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This fixes all direct leaks found by running `ctest -R deterministic-unit_test_wavefunction_trialwf`. When possible, I used `std::unique_ptr` to make ownership more clear.

This is part of #3083

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Ubuntu 20.04

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- N/A. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- N/A. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
1. could you squash current commits to keep the history clean?
2. place fix legacy_cuda deterministic tests.

**Review by ye-luo — CHANGES_REQUESTED** [review]
See my comments above

**Comment by quantumsteve** [comment]
> 2\. place fix legacy_cuda deterministic tests.

Would you enable the other CI builds so I can see the test failure?

**Comment by ye-luo** [comment]
> > 2. place fix legacy_cuda deterministic tests.
> 
> Would you enable the other CI builds so I can see the test failure?

We don't have a CI running on GPU, the current code compiles but fails at run.

**Comment by quantumsteve** [comment]
I reduced the change causing the test failure to these 4 lines.

```
diff --git a/src/QMCWaveFunctions/EinsplineSetBuilder_createSPOs.cpp b/src/QMCWaveFunctions/EinsplineSetBuilder_createSPOs.cpp
index 95464c806..cd05b6c23 100644
--- a/src/QMCWaveFunctions/EinsplineSetBuilder_createSPOs.cpp
+++ b/src/QMCWaveFunctions/EinsplineSetBuilder_createSPOs.cpp
@@ -336,7 +336,7 @@ std::unique_ptr<SPOSet> EinsplineSetBuilder::createSPOSetFromXML(xmlNodePtr cur)
   auto bspline_zd = MixedSplineReader->create_spline_set(spinSet, spo_cur);
   if (!bspline_zd)
     APP_ABORT_TRACE(__FILE__, __LINE__, "Failed to create SPOSet*");
-  auto OrbitalSet = bspline_zd.release();
+  auto OrbitalSet = std::move(bspline_zd);
 #if defined(MIXED_PRECISION)
   if (use_einspline_set_extended == "yes")
   {
@@ -362,7 +362,7 @@ std::unique_ptr<SPOSet> EinsplineSetBuilder::createSPOSetFromXML(xmlNodePtr cur)
       temp_OrbitalSet->MultiSpline->num_splines = NumDistinctOrbitals;
       temp_OrbitalSet->resizeStorage(NumDistinctOrbitals, NumValenceOrbs);
       //set the flags for anti periodic boundary conditions
-      temp_OrbitalSet->HalfG = dynamic_cast<BsplineSet*>(OrbitalSet)->getHalfG();
+      temp_OrbitalSet->HalfG = dynamic_cast<BsplineSet*>(OrbitalSet.get())->getHalfG();
       new_OrbitalSet         = temp_OrbitalSet;
     }
     else
@@ -388,7 +388,7 @@ std::unique_ptr<SPOSet> EinsplineSetBuilder::createSPOSetFromXML(xmlNodePtr cur)
     }
     //set the internal parameters
     setTiling(new_OrbitalSet, numOrbs);
-    OrbitalSet = new_OrbitalSet;
+    OrbitalSet.reset(new_OrbitalSet);
   }
 #endif
   app_log() << "Time spent in creating B-spline SPOs " << mytimer.elapsed() << "sec" << std::endl;
@@ -472,8 +472,8 @@ std::unique_ptr<SPOSet> EinsplineSetBuilder::createSPOSetFromXML(xmlNodePtr cur)
   }
 #endif
   OrbitalSet->finalizeConstruction();
-  SPOSetMap[aset] = OrbitalSet;
-  return std::unique_ptr<SPOSet>(OrbitalSet);
+  SPOSetMap[aset] = OrbitalSet.get();
+  return OrbitalSet;
 }
 
 std::unique_ptr<SPOSet> EinsplineSetBuilder::createSPOSet(xmlNodePtr cur, SPOSetInputInfo& input_info)
 ```
 
 I wonder if deleting `bspline_zd` is somehow causing the segfault. Any ideas?

**Comment by prckent** [comment]
Perhaps you have already verified this, but builds with QMC_COMPLEX=1 &/or QMC_MIXED_PRECISION=0 could be revealing.

File: src/QMCWaveFunctions/LCAO/SoaAtomicBasisSet.h [hunk 1]
```diff
@@ -61,14 +61,26 @@ struct SoaAtomicBasisSet
...
+        tempS(in.tempS)
+  {}
+
+  ~SoaAtomicBasisSet() = default;
 
   SoaAtomicBasisSet<ROT, SH>* makeClone() const
   {
     SoaAtomicBasisSet<ROT, SH>* myclone = new SoaAtomicBasisSet<ROT, SH>(*this);
-    myclone->MultiRnl                   = MultiRnl->makeClone();
+    myclone->MultiRnl.reset(MultiRnl->makeClone());
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/SoaAtomicBasisSet.h:83]: The copy constructor should have handled this.

File: src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.h [hunk 2]
```diff
@@ -337,7 +350,7 @@ struct SoaLocalizedBasisSet : public SoaBasisSetBase<ORBT>
    * @param icenter the index of the center
    * @param aos a set of Centered Atomic Orbitals
    */
-  void add(int icenter, COT* aos) { LOBasisSet[icenter] = aos; }
+  void add(int icenter, std::unique_ptr<COT> aos) { LOBasisSet[icenter] = std::move(aos); }
```
Comment (ye-luo) [inline src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.h:353]: I think rvalue reference is better.

File: src/QMCWaveFunctions/SPOSetBuilder.cpp [hunk 3]
```diff
@@ -86,7 +86,7 @@ SPOSet* SPOSetBuilder::createSPOSet(xmlNodePtr cur)
     abort();
 #else
     // create sposet with rotation
-    auto* rot_spo   = new RotatedSPOs(sposet);
+    auto rot_spo    = std::make_unique<RotatedSPOs>(sposet.get());
```
Comment (ye-luo) [inline src/QMCWaveFunctions/SPOSetBuilder.cpp:89]: please change RotatedSPOs::Phi to unique_ptr as well as the constructor

Comment (ye-luo) [reply src/QMCWaveFunctions/SPOSetBuilder.cpp]: fixed

**Comment by ye-luo** [comment]
Test this please

File: src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.h (hunk 2 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/LCAO/SoaLocalizedBasisSet.h:353]: value is better than rvalue for less ambiguity.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
@quantumsteve and @williamfgc I'm looking at the LCAO to redefine pointer ownership + simplfy the code. Leave this to me right now and feel free to touch Jastrow.

**Comment by quantumsteve** [comment]
@ye-luo Do the `deterministic-unit_test_hamiltonian_coulomb`, `deterministic-unit_test_hamiltonian_force` or `deterministic-unit_test_hamiltonian_ham` tests touch the LCAO code you're working on?

**Comment by prckent** [comment]
@quantumsteve  https://github.com/QMCPACK/qmcpack/pull/3223

**Comment by ye-luo** [comment]
@quantumsteve I don't have pending working in LCAO.

**Pull request merged**